File: ios-native/RemittanceApp/Services/SearchService.swift
import Foundation
import Combine

// MARK: - Search Index Types
enum SearchIndex: String, Codable, CaseIterable {
    case transactions
    case users
    case beneficiaries
    case disputes
    case auditLogs = "audit_logs"
    case kyc
    case wallets
    case cards
    case bills
    case airtime
}

// MARK: - Search Request Models
struct SearchQuery: Codable {
    let query: String
    let index: [String]?
    let filters: [String: String]?
    let sort: SearchSort?
    let pagination: SearchPagination?
    let highlight: Bool
    let aggregations: [String]?
    
    init(
        query: String,
        index: [SearchIndex]? = nil,
        filters: [String: String]? = nil,
        sort: SearchSort? = nil,
        pagination: SearchPagination? = nil,
        highlight: Bool = true,
        aggregations: [String]? = nil
    ) {
        self.query = query
        self.index = index?.map { $0.rawValue }
        self.filters = filters
        self.sort = sort
        self.pagination = pagination
        self.highlight = highlight
        self.aggregations = aggregations
    }
}

struct SearchSort: Codable {
    let field: String
    let order: String
    
    init(field: String, order: String = "desc") {
        self.field = field
        self.order = order
    }
}

struct SearchPagination: Codable {
    let page: Int
    let size: Int
    
    init(page: Int = 1, size: Int = 20) {
        self.page = page
        self.size = size
    }
}

// MARK: - Search Response Models
struct SearchResponse<T: Codable>: Codable {
    let hits: [SearchHit<T>]
    let total: Int
    let page: Int
    let size: Int
    let took: Int
    let aggregations: [String: [AggregationBucket]]?
}

struct SearchHit<T: Codable>: Codable {
    let id: String
    let index: String
    let score: Float
    let source: T
    let highlight: [String: [String]]?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case index = "_index"
        case score = "_score"
        case source = "_source"
        case highlight
    }
}

struct AggregationBucket: Codable {
    let key: String
    let count: Int
    
    enum CodingKeys: String, CodingKey {
        case key
        case count = "doc_count"
    }
}

// MARK: - Domain-specific Result Types
struct TransactionSearchResult: Codable, Identifiable {
    let id: String
    let reference: String
    let type: String
    let amount: Double
    let currency: String
    let status: String
    let description: String
    let createdAt: String
    let senderId: String?
    let recipientId: String?
    
    enum CodingKeys: String, CodingKey {
        case id, reference, type, amount, currency, status, description
        case createdAt = "created_at"
        case senderId = "sender_id"
        case recipientId = "recipient_id"
    }
}

struct BeneficiarySearchResult: Codable, Identifiable {
    let id: String
    let name: String
    let accountNumber: String
    let bankCode: String
    let bankName: String
    let country: String
    let currency: String
    let createdAt: String
    
    enum CodingKeys: String, CodingKey {
        case id, name, country, currency
        case accountNumber = "account_number"
        case bankCode = "bank_code"
        case bankName = "bank_name"
        case createdAt = "created_at"
    }
}

struct DisputeSearchResult: Codable, Identifiable {
    let id: String
    let transactionId: String
    let type: String
    let status: String
    let description: String
    let createdAt: String
    let resolvedAt: String?
    
    enum CodingKeys: String, CodingKey {
        case id, type, status, description
        case transactionId = "transaction_id"
        case createdAt = "created_at"
        case resolvedAt = "resolved_at"
    }
}

struct AuditLogSearchResult: Codable, Identifiable {
    let id: String
    let action: String
    let category: String
    let userId: String
    let resourceType: String
    let resourceId: String
    let details: String
    let ipAddress: String
    let timestamp: String
    
    enum CodingKeys: String, CodingKey {
        case id, action, category, details, timestamp
        case userId = "user_id"
        case resourceType = "resource_type"
        case resourceId = "resource_id"
        case ipAddress = "ip_address"
    }
}

struct SearchSuggestion: Codable, Identifiable {
    var id: String { text }
    let text: String
    let score: Float
    let index: String
}

struct RecentSearch: Codable, Identifiable {
    var id: String { query + (index ?? "") }
    let query: String
    let index: String?
    let timestamp: String
}

// MARK: - Search Service
class SearchService: ObservableObject {
    static let shared = SearchService()
    
    private let baseURL: String
    private var authToken: String?
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder
    
    @Published var isLoading = false
    @Published var error: Error?
    
    init(
        baseURL: String = "https://api.remittance.com/api/search",
        authToken: String? = nil
    ) {
        self.baseURL = baseURL
        self.authToken = authToken
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: config)
        
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }
    
    func setAuthToken(_ token: String) {
        self.authToken = token
    }
    
    // MARK: - Unified Search
    func search<T: Codable>(query: SearchQuery) async throws -> SearchResponse<T> {
        let url = URL(string: "\(baseURL)/unified")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(query)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode(SearchResponse<T>.self, from: data)
    }
    
    // MARK: - Transaction Search
    func searchTransactions(
        query: String,
        filters: [String: String]? = nil,
        pagination: SearchPagination = SearchPagination()
    ) async throws -> SearchResponse<TransactionSearchResult> {
        let searchQuery = SearchQuery(
            query: query,
            index: [.transactions],
            filters: filters,
            pagination: pagination
        )
        
        let url = URL(string: "\(baseURL)/transactions")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(searchQuery)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode(SearchResponse<TransactionSearchResult>.self, from: data)
    }
    
    // MARK: - Beneficiary Search
    func searchBeneficiaries(
        query: String,
        filters: [String: String]? = nil,
        pagination: SearchPagination = SearchPagination()
    ) async throws -> SearchResponse<BeneficiarySearchResult> {
        let searchQuery = SearchQuery(
            query: query,
            index: [.beneficiaries],
            filters: filters,
            pagination: pagination
        )
        
        let url = URL(string: "\(baseURL)/beneficiaries")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(searchQuery)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode(SearchResponse<BeneficiarySearchResult>.self, from: data)
    }
    
    // MARK: - Dispute Search
    func searchDisputes(
        query: String,
        filters: [String: String]? = nil,
        pagination: SearchPagination = SearchPagination()
    ) async throws -> SearchResponse<DisputeSearchResult> {
        let searchQuery = SearchQuery(
            query: query,
            index: [.disputes],
            filters: filters,
            pagination: pagination
        )
        
        let url = URL(string: "\(baseURL)/disputes")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(searchQuery)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode(SearchResponse<DisputeSearchResult>.self, from: data)
    }
    
    // MARK: - Audit Log Search
    func searchAuditLogs(
        query: String,
        filters: [String: String]? = nil,
        pagination: SearchPagination = SearchPagination()
    ) async throws -> SearchResponse<AuditLogSearchResult> {
        let searchQuery = SearchQuery(
            query: query,
            index: [.auditLogs],
            filters: filters,
            pagination: pagination
        )
        
        let url = URL(string: "\(baseURL)/audit-logs")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(searchQuery)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode(SearchResponse<AuditLogSearchResult>.self, from: data)
    }
    
    // MARK: - Suggestions
    func getSuggestions(query: String, index: SearchIndex? = nil) async throws -> [SearchSuggestion] {
        var urlComponents = URLComponents(string: "\(baseURL)/suggestions")!
        urlComponents.queryItems = [URLQueryItem(name: "q", value: query)]
        if let index = index {
            urlComponents.queryItems?.append(URLQueryItem(name: "index", value: index.rawValue))
        }
        
        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = "GET"
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode([SearchSuggestion].self, from: data)
    }
    
    // MARK: - Recent Searches
    func getRecentSearches() async throws -> [RecentSearch] {
        let url = URL(string: "\(baseURL)/recent")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
        
        return try decoder.decode([RecentSearch].self, from: data)
    }
    
    func saveRecentSearch(query: String, index: SearchIndex? = nil) async throws {
        let url = URL(string: "\(baseURL)/recent")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        var body: [String: String] = ["query": query]
        if let index = index {
            body["index"] = index.rawValue
        }
        request.httpBody = try encoder.encode(body)
        
        let (_, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
    }
    
    func clearRecentSearches() async throws {
        let url = URL(string: "\(baseURL)/recent")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        let (_, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SearchError.requestFailed
        }
    }
}

// MARK: - Search Errors
enum SearchError: Error, LocalizedError {
    case requestFailed
    case invalidResponse
    case decodingFailed
    case networkUnavailable
    
    var errorDescription: String? {
        switch self {
        case .requestFailed:
            return "Search request failed"
        case .invalidResponse:
            return "Invalid response from server"
        case .decodingFailed:
            return "Failed to decode search results"
        case .networkUnavailable:
            return "Network unavailable"
        }
    }
}

File: android-native/app/src/main/kotlin/com/remittance/app/data/remote/SearchService.kt
package com.pos54link.app.data.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * OpenSearch Integration Service for Android Native App
 * Connects to the unified search service endpoints
 */

// Search Index Types
enum class SearchIndex(val value: String) {
    TRANSACTIONS("transactions"),
    USERS("users"),
    BENEFICIARIES("beneficiaries"),
    DISPUTES("disputes"),
    AUDIT_LOGS("audit_logs"),
    KYC("kyc"),
    WALLETS("wallets"),
    CARDS("cards"),
    BILLS("bills"),
    AIRTIME("airtime")
}

// Search Request Models
@Serializable
data class SearchQuery(
    val query: String,
    val index: List<String>? = null,
    val filters: Map<String, String>? = null,
    val sort: SearchSort? = null,
    val pagination: SearchPagination? = null,
    val highlight: Boolean = true,
    val aggregations: List<String>? = null
)

@Serializable
data class SearchSort(
    val field: String,
    val order: String = "desc"
)

@Serializable
data class SearchPagination(
    val page: Int = 1,
    val size: Int = 20
)

// Search Response Models
@Serializable
data class SearchResponse<T>(
    val hits: List<SearchHit<T>>,
    val total: Int,
    val page: Int,
    val size: Int,
    val took: Long,
    val aggregations: Map<String, List<AggregationBucket>>? = null
)

@Serializable
data class SearchHit<T>(
    val id: String,
    val index: String,
    val score: Float,
    val source: T,
    val highlight: Map<String, List<String>>? = null
)

@Serializable
data class AggregationBucket(
    val key: String,
    val count: Int
)

// Domain-specific result types
@Serializable
data class TransactionSearchResult(
    val id: String,
    val reference: String,
    val type: String,
    val amount: Double,
    val currency: String,
    val status: String,
    val description: String,
    val createdAt: String,
    val senderId: String? = null,
    val recipientId: String? = null
)

@Serializable
data class BeneficiarySearchResult(
    val id: String,
    val name: String,
    val accountNumber: String,
    val bankCode: String,
    val bankName: String,
    val country: String,
    val currency: String,
    val createdAt: String
)

@Serializable
data class DisputeSearchResult(
    val id: String,
    val transactionId: String,
    val type: String,
    val status: String,
    val description: String,
    val createdAt: String,
    val resolvedAt: String? = null
)

@Serializable
data class AuditLogSearchResult(
    val id: String,
    val action: String,
    val category: String,
    val userId: String,
    val resourceType: String,
    val resourceId: String,
    val details: String,
    val ipAddress: String,
    val timestamp: String
)

@Serializable
data class SearchSuggestion(
    val text: String,
    val score: Float,
    val index: String
)

@Serializable
data class RecentSearch(
    val query: String,
    val index: String? = null,
    val timestamp: String
)

/**
 * OpenSearch Service Implementation
 */
class SearchService(
    private val baseUrl: String = "https://api.remittance.com/api/search",
    private val authToken: String? = null
) {
    private val json = Json { 
        ignoreUnknownKeys = true 
        isLenient = true
    }
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val mediaType = "application/json; charset=utf-8".toMediaType()

    /**
     * Unified search across all indices
     */
    suspend fun search(query: SearchQuery): Result<SearchResponse<Map<String, Any>>> {
        return withContext(Dispatchers.IO) {
            try {
                val requestBody = json.encodeToString(SearchQuery.serializer(), query)
                    .toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/unified")
                    .post(requestBody)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    // Parse response - simplified for demonstration
                    Result.success(parseSearchResponse(body))
                } else {
                    Result.failure(Exception("Search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Search transactions
     */
    suspend fun searchTransactions(
        query: String,
        filters: Map<String, String>? = null,
        pagination: SearchPagination = SearchPagination()
    ): Result<SearchResponse<TransactionSearchResult>> {
        return withContext(Dispatchers.IO) {
            try {
                val searchQuery = SearchQuery(
                    query = query,
                    index = listOf(SearchIndex.TRANSACTIONS.value),
                    filters = filters,
                    pagination = pagination,
                    highlight = true
                )
                
                val requestBody = json.encodeToString(SearchQuery.serializer(), searchQuery)
                    .toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/transactions")
                    .post(requestBody)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    Result.success(parseTransactionResponse(body))
                } else {
                    Result.failure(Exception("Transaction search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Search beneficiaries
     */
    suspend fun searchBeneficiaries(
        query: String,
        filters: Map<String, String>? = null,
        pagination: SearchPagination = SearchPagination()
    ): Result<SearchResponse<BeneficiarySearchResult>> {
        return withContext(Dispatchers.IO) {
            try {
                val searchQuery = SearchQuery(
                    query = query,
                    index = listOf(SearchIndex.BENEFICIARIES.value),
                    filters = filters,
                    pagination = pagination,
                    highlight = true
                )
                
                val requestBody = json.encodeToString(SearchQuery.serializer(), searchQuery)
                    .toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/beneficiaries")
                    .post(requestBody)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    Result.success(parseBeneficiaryResponse(body))
                } else {
                    Result.failure(Exception("Beneficiary search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Search disputes
     */
    suspend fun searchDisputes(
        query: String,
        filters: Map<String, String>? = null,
        pagination: SearchPagination = SearchPagination()
    ): Result<SearchResponse<DisputeSearchResult>> {
        return withContext(Dispatchers.IO) {
            try {
                val searchQuery = SearchQuery(
                    query = query,
                    index = listOf(SearchIndex.DISPUTES.value),
                    filters = filters,
                    pagination = pagination,
                    highlight = true
                )
                
                val requestBody = json.encodeToString(SearchQuery.serializer(), searchQuery)
                    .toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/disputes")
                    .post(requestBody)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    Result.success(parseDisputeResponse(body))
                } else {
                    Result.failure(Exception("Dispute search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Search audit logs
     */
    suspend fun searchAuditLogs(
        query: String,
        filters: Map<String, String>? = null,
        pagination: SearchPagination = SearchPagination()
    ): Result<SearchResponse<AuditLogSearchResult>> {
        return withContext(Dispatchers.IO) {
            try {
                val searchQuery = SearchQuery(
                    query = query,
                    index = listOf(SearchIndex.AUDIT_LOGS.value),
                    filters = filters,
                    pagination = pagination,
                    highlight = true
                )
                
                val requestBody = json.encodeToString(SearchQuery.serializer(), searchQuery)
                    .toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/audit-logs")
                    .post(requestBody)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "{}"
                    Result.success(parseAuditLogResponse(body))
                } else {
                    Result.failure(Exception("Audit log search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Get search suggestions (autocomplete)
     */
    suspend fun getSuggestions(
        query: String,
        index: SearchIndex? = null
    ): Result<List<SearchSuggestion>> {
        return withContext(Dispatchers.IO) {
            try {
                val url = buildString {
                    append("$baseUrl/suggestions?q=$query")
                    index?.let { append("&index=${it.value}") }
                }
                
                val request = Request.Builder()
                    .url(url)
                    .get()
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "[]"
                    Result.success(parseSuggestions(body))
                } else {
                    Result.failure(Exception("Suggestions failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Get recent searches
     */
    suspend fun getRecentSearches(): Result<List<RecentSearch>> {
        return withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$baseUrl/recent")
                    .get()
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    val body = response.body?.string() ?: "[]"
                    Result.success(parseRecentSearches(body))
                } else {
                    Result.failure(Exception("Recent searches failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Save a recent search
     */
    suspend fun saveRecentSearch(query: String, index: SearchIndex? = null): Result<Unit> {
        return withContext(Dispatchers.IO) {
            try {
                val body = buildString {
                    append("{\"query\":\"$query\"")
                    index?.let { append(",\"index\":\"${it.value}\"") }
                    append("}")
                }.toRequestBody(mediaType)
                
                val request = Request.Builder()
                    .url("$baseUrl/recent")
                    .post(body)
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                        addHeader("Content-Type", "application/json")
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    Result.success(Unit)
                } else {
                    Result.failure(Exception("Save recent search failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    /**
     * Clear recent searches
     */
    suspend fun clearRecentSearches(): Result<Unit> {
        return withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$baseUrl/recent")
                    .delete()
                    .apply {
                        authToken?.let { addHeader("Authorization", "Bearer $it") }
                    }
                    .build()
                
                val response = client.newCall(request).execute()
                
                if (response.isSuccessful) {
                    Result.success(Unit)
                } else {
                    Result.failure(Exception("Clear recent searches failed: ${response.code}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
    }

    // Response parsing helpers
    private fun parseSearchResponse(body: String): SearchResponse<Map<String, Any>> {
        // Simplified parsing - in production use proper JSON deserialization
        return SearchResponse(
            hits = emptyList(),
            total = 0,
            page = 1,
            size = 20,
            took = 0
        )
    }

    private fun parseTransactionResponse(body: String): SearchResponse<TransactionSearchResult> {
        return SearchResponse(
            hits = emptyList(),
            total = 0,
            page = 1,
            size = 20,
            took = 0
        )
    }

    private fun parseBeneficiaryResponse(body: String): SearchResponse<BeneficiarySearchResult> {
        return SearchResponse(
            hits = emptyList(),
            total = 0,
            page = 1,
            size = 20,
            took = 0
        )
    }

    private fun parseDisputeResponse(body: String): SearchResponse<DisputeSearchResult> {
        return SearchResponse(
            hits = emptyList(),
            total = 0,
            page = 1,
            size = 20,
            took = 0
        )
    }

    private fun parseAuditLogResponse(body: String): SearchResponse<AuditLogSearchResult> {
        return SearchResponse(
            hits = emptyList(),
            total = 0,
            page = 1,
            size = 20,
            took = 0
        )
    }

    private fun parseSuggestions(body: String): List<SearchSuggestion> {
        return emptyList()
    }

    private fun parseRecentSearches(body: String): List<RecentSearch> {
        return emptyList()
    }

    companion object {
        @Volatile
        private var instance: SearchService? = null

        fun getInstance(baseUrl: String? = null, authToken: String? = null): SearchService {
            return instance ?: synchronized(this) {
                instance ?: SearchService(
                    baseUrl = baseUrl ?: "https://api.remittance.com/api/search",
                    authToken = authToken
                ).also { instance = it }
            }
        }
    }
}
